jax/_src/numpy/lax_numpy.py

```diff
@@ -1368,6 +1368,87 @@ def gradient(f, *varargs, axis: Optional[Union[int, Tuple[int, ...]]] = None,
 def isrealobj(x):
   return not iscomplexobj(x)
 
+@partial(jit, static_argnums=1)
+class _RankWarning(UserWarning):
+    """
+    Issued by `polyfit` when the Vandermonde matrix is rank deficient.
+    For more information, a way to suppress the warning, and an example of
+    `RankWarning` being issued, see `polyfit`.
+    """
+    pass
+
+@_wraps(np.polyfit)
+def polyfit(x, y, deg, rcond=None, full=False, w=None, cov=False):
+  _check_arraylike("polyfit", x, y, deg)
+  x = core.concrete_or_error(None, x, "The error arose in polyfit()")
+  y = core.concrete_or_error(None, y, "The error arose in polyfit()")
+  order = int(deg) + 1
+  # check arguments.
+  if deg < 0:
+      raise ValueError("expected deg >= 0")
+  if x.ndim != 1:
+      raise TypeError("expected 1D vector for x")
+  if x.size == 0:
+      raise TypeError("expected non-empty vector for x")
+  if y.ndim < 1 or y.ndim > 2:
+      raise TypeError("expected 1D or 2D array for y")
+  if x.shape[0] != y.shape[0]:
+      raise TypeError("expected x and y to have same length")
+
+  # set rcond
+  if rcond is None:
+      rcond = len(x)*finfo(x.dtype).eps
+
+  # set up least squares equation for powers of x
+  lhs = vander(x, order)
+  rhs = y
+
+  # apply weighting
+  if w is not None:
+      w = core.concrete_or_error(None, w, "The error arose in polyfit()") + 0.0
```

Comment:
I don't think `w` needs to be concrete, so I would remove this.

It looks like it does need to be arraylike, so `_check_arraylike` would be applicable.

If you need to ensure it's floating point, use `_promote_dtypes_inexact` rather than adding 0.0.